Drag & Drop Navigation Reordering

PaulAdamDavis opened the pull request:
Closes #4540

- Implements drag & drop to reorder navigation items
- Adds a `order` property to navigation items
- Adds a tiny library to enable touch events for drag & drop. It hooks onto jQuery UI.
- Sort nav items before being saved
- Adds `settings-view-navigation` to route for body class

ErisDS commented:
I just took this for a spin, and it seems like the drag & drop works fine, but adding a new item and pressing save sends that item to position 2 in the list rather than the last position.

PaulAdamDavis commented:
Updated, almost complete reworked actually. I've added a `sort` attribute to the JSON stored in the DB.

There's one little bug, but this is very close I think.

PaulAdamDavis commented:
Updated again, and I think I've solved any issues. Here's a gif that shows reordering, and adding multiple news items and reordering, then saving at the end. The order remains.

Because there's now a `sort` attribute stored in the JSON, we should either sort nav items in the `{{nav}}` theme helper, or store nav items in the correct order when saving to the database. Thoughts welcome. :)

![nav-drag-drop](https://cloud.githubusercontent.com/assets/390392/6065532/fd2f850c-ad5d-11e4-9949-183085ee91eb.gif)

nuclearpengy commented:
Looks rad! \o/

PaulAdamDavis commented:
Updated again to sort nav items before being saved.

novaugust commented on core/client/components/gh-navitem.js:
```diff
@@ -1,5 +1,11 @@
+    attributeBindings:['getSort:data-sort'],
+    getSort: function () {
+        return this.get('navItem.sort');
+    }.property('navItem.sort'),
```
We moved away from `property` in favor of `Ember.computed`, if you would my good man :D You can actually make this sort of behaviour really simple by doing
```js
attributeBindings: ['sort:data-sort'],
sort: Ember.computed.readOnly('navItem.sort')
```

novaugust commented on core/client/controllers/settings/navigation.js:
```diff
@@ -31,6 +32,11 @@ NavigationController = Ember.Controller.extend({
+        // Order by `sort` attribute
+        navItems.sort(function (a, b) {
+            return a.get('sort') - b.get('sort');
+        });
```
Ember's got a convenient method for sorting by attributes
```js
navItems.sortBy('sort')
```

novaugust commented on core/client/views/settings/navigation.js:
```diff
@@ -1,6 +1,28 @@
+
+    didInsertElement: function () {
+        var controller = this.get('controller');
+        Ember.$('#settings-navigation').sortable({
```
might want to `.js-sortable-navigation` that selector

novaugust commented on core/client/assets/sass/layouts/settings.scss:
```diff
@@ -394,18 +394,29 @@
     @media (min-width: 601px) {
         margin-bottom: 10px;
+
+        &.last-navigation-item {
```
Why don't you use the `:last-child` pseudo-class to style / jquery select this instead of having to add another class? I think it'd clean up the javascript a bit

novaugust commented on core/client/views/settings/navigation.js:
```diff
@@ -1,6 +1,28 @@
+
+            update: function () {
+                var indexes = [];
+                Ember.$(this).find('.navigation-item:not(.last-navigation-item)').each(function () {
```
is the `this` here the `#settings-navigation` element?

novaugust commented:
Ember looks fine with me, talked with paul in chat a little about how his jquery is similar to a pattern used in our tab panes for parent/child view interactions.

My only other comment is, can `sort` be renamed to `order` or something along those lines?
